File: agents-node/apiServer.js
import express from "express";
import cors from "cors";
import OpenAI from "openai";
import { AGENT_PROFILES } from "./profiles.js";
import { AGENT_KEYS, OPENAI_API_KEY, CHAINS } from "./config.js";
import { ReadOnlyMarketClient } from "./marketClient.js";
import { ethers } from "ethers";

let _marketClient = null; // default (base) client for orchestrator
let _orchestrator = null;

// Read-only clients per chain (lazy initialized)
const _chainClients = {};

function getChainClient(chainKey) {
  // If it's the default chain and we have the orchestrator's client, use that
  if (chainKey === "base" && _marketClient) return _marketClient;

  if (!_chainClients[chainKey]) {
    const cfg = CHAINS[chainKey];
    if (!cfg || !cfg.contractAddress) return null;
    _chainClients[chainKey] = new ReadOnlyMarketClient(cfg.rpcUrl, cfg.contractAddress);
  }
  return _chainClients[chainKey];
}

function resolveClient(req) {
  const chain = req.query.chain || "base";
  const client = getChainClient(chain);
  return { client, chain };
}

export function setSharedState(marketClient, orchestrator) {
  _marketClient = marketClient;
  _orchestrator = orchestrator;
}

function getAgentNameMap() {
  const mapping = {};
  for (let i = 0; i < AGENT_PROFILES.length; i++) {
    const key = AGENT_KEYS[i];
    if (key) {
      try {
        const wallet = new ethers.Wallet(key);
        mapping[wallet.address.toLowerCase()] = AGENT_PROFILES[i].name;
      } catch {}
    }
  }
  return mapping;
}

export function startApiServer(port = 8000) {
  const app = express();
  app.use(cors());
  app.use(express.json());

  // Initialize monad client eagerly so it's ready
  getChainClient("monad");

  const statsCache = { data: null, ts: 0 };
  app.get("/api/stats", async (req, res) => {
    const now = Date.now();
    if (statsCache.data && now - statsCache.ts < 60000) {
      return res.json(statsCache.data);
    }
    const { client } = resolveClient(req);
    if (!client) return res.json({ total_agents: 0, total_markets: 0 });
    try {
      const marketCount = await client.getMarketCount();
      const total_agents = AGENT_PROFILES.length;
      const result = { total_agents, total_markets: marketCount };
      statsCache.data = result;
      statsCache.ts = now;
      res.json(result);
    } catch (e) {
      res.status(500).json({ detail: e.message });
    }
  });

  app.get("/api/health", (req, res) => {
    res.json({
      status: "ok",
      mode: _orchestrator ? "orchestrate" : "standalone",
      market_client: !!_marketClient,
      orchestrator: !!_orchestrator,
      chains: Object.keys(CHAINS),
    });
  });

  app.get("/api/chains", (req, res) => {
    res.json(
      Object.entries(CHAINS).map(([key, cfg]) => ({
        key,
        name: cfg.name,
        chainId: cfg.chainId,
        symbol: cfg.symbol,
        active: !!cfg.contractAddress,
      }))
    );
  });

  app.get("/api/markets", async (req, res) => {
    const { client } = resolveClient(req);
    if (!client) return res.status(503).json({ detail: "Chain not initialized" });
    try {
      const count = await client.getMarketCount();
      const markets = [];
      for (let i = 0; i < count; i++) {
        try {
          const [info, active] = await Promise.all([
            client.getMarketInfo(i),
            client.isMarketActive(i),
          ]);
          markets.push({
            market_id: i,
            question: info.question,
            creator: info.creator,
            current_price: Number(info.currentPrice) / 1e18,
            resolved: info.resolved,
            is_active: active,
            prediction_count: info.predictionCount,
            total_pool: Number(info.totalPool) / 1e18,
          });
        } catch {}
      }
      res.json(markets);
    } catch (e) {
      res.status(500).json({ detail: e.message });
    }
  });

  app.get("/api/markets/count", async (req, res) => {
    const { client } = resolveClient(req);
    if (!client) return res.status(503).json({ detail: "Chain not initialized" });
    try {
      const count = await client.getMarketCount();
      res.json({ count });
    } catch (e) {
      res.status(500).json({ detail: e.message });
    }
  });

  app.get("/api/markets/:id", async (req, res) => {
    const { client } = resolveClient(req);
    if (!client) return res.status(503).json({ detail: "Chain not initialized" });
    const marketId = parseInt(req.params.id);
    try {
      const info = await client.getMarketInfo(marketId);
      const params = await client.getMarketParams(marketId);
      const predictions = await client.getPredictions(marketId);
      res.json({
        market_id: marketId,
        question: info.question,
        current_price: Number(info.currentPrice) / 1e18,
        creator: info.creator,
        resolved: info.resolved,
        total_pool: Number(info.totalPool) / 1e18,
        prediction_count: info.predictionCount,
        params: {
          alpha: Number(params.alpha) / 1e18,
          k: params.k,
          flat_reward: Number(params.flatReward) / 1e18,
          bond_amount: Number(params.bondAmount) / 1e18,
          liquidity_param: Number(params.liquidityParam) / 1e18,
          created_at: params.createdAt,
        },
        predictions: predictions.map((p, i) => ({
          index: i,
          predictor: p.predictor,
          probability: Number(p.probability) / 1e18,
          price_before: Number(p.priceBefore) / 1e18,
          price_after: Number(p.priceAfter) / 1e18,
          bond: Number(p.bond) / 1e18,
          timestamp: p.timestamp,
        })),
      });
    } catch (e) {
      res.status(404).json({ detail: `Market not found: ${e.message}` });
    }
  });

  app.get("/api/leaderboard", async (req, res) => {
    const { client } = resolveClient(req);
    if (!client) return res.json({ rankings: [] });
    try {
      const nameMap = getAgentNameMap();
      const agentAddresses = Object.keys(nameMap);
      if (!agentAddresses.length) return res.json({ rankings: [] });

      const marketCount = await client.getMarketCount();
      const totals = {};
      for (const addr of agentAddresses) {
        totals[addr] = 0;
      }

      for (let i = 0; i < marketCount; i++) {
        const info = await client.getMarketInfo(i);
        if (!info.resolved) continue;

        await Promise.all(agentAddresses.map(async (addr) => {
          try {
            const predicted = await client.hasPredicted(i, addr);
            if (!predicted) return;
            const payout = await client.getPayout(i, addr);
            const params = await client.getMarketParams(i);
            const net = Number(payout) / 1e18 - Number(params.bondAmount) / 1e18;
            totals[addr] += net;
          } catch {}
        }));
      }

      const sorted = Object.entries(totals).sort((a, b) => b[1] - a[1]);
      res.json({
        rankings: sorted.map(([addr, eth], i) => ({
          rank: i + 1,
          agent: nameMap[addr] || addr,
          total_eth: Math.round(eth * 1e6) / 1e6,
        })),
      });
    } catch (e) {
      res.status(500).json({ detail: e.message });
    }
  });

  app.get("/api/markets/:id/leaderboard", async (req, res) => {
    const { client } = resolveClient(req);
    if (!client) return res.json({ rankings: [] });
    const marketId = parseInt(req.params.id);
    try {
      const info = await client.getMarketInfo(marketId);
      if (!info.resolved) return res.json({ rankings: [] });

      const nameMap = getAgentNameMap();
      const predictions = await client.getPredictions(marketId);
      const params = await client.getMarketParams(marketId);
      const bondAmt = Number(params.bondAmount) / 1e18;

      const rankings = await Promise.all(predictions.map(async (p, i) => {
        const addr = p.predictor.toLowerCase();
        try {
          const payout = await client.getPayout(marketId, p.predictor);
          const net = Number(payout) / 1e18 - bondAmt;
          return {
            rank: 0,
            agent: nameMap[addr] || `${p.predictor.slice(0, 6)}...${p.predictor.slice(-4)}`,
            address: p.predictor,
            total_eth: Math.round(net * 1e6) / 1e6,
            probability: Number(p.probability) / 1e18,
          };
        } catch {
          return null;
        }
      }));

      const valid = rankings.filter(Boolean).sort((a, b) => b.total_eth - a.total_eth);
      valid.forEach((r, i) => r.rank = i + 1);

      res.json({ rankings: valid });
    } catch (e) {
      res.status(500).json({ detail: e.message });
    }
  });

  app.post("/api/validate-question", async (req, res) => {
    const { question } = req.body;
    if (!question || question.trim().length < 10) {
      return res.json({ valid: false, reason: "Question is too short. Please provide a more detailed question." });
    }

    try {
      const openai = new OpenAI({ apiKey: OPENAI_API_KEY });
      const response = await openai.chat.completions.create({
        model: "gpt-4o-mini",
        messages: [
          {
            role: "system",
            content: `You are a question validator for a self-resolving prediction market (based on the paper "Self-Resolving Prediction Markets for Unverifiable Outcomes" by Yiling Chen et al.).

This market is designed for questions that are either:
1. UNVERIFIABLE — No oracle, API, or data source can definitively determine the true answer. These are subjective, philosophical, or opinion-based questions.
2. LONG-HORIZON — The outcome is so far in the future (years/decades) that no current oracle can resolve it, and locking capital until resolution is impractical.

REJECT questions that are:
- Short-term verifiable facts (e.g., "Will ETH hit $5000 tomorrow?" — a price oracle can verify this)
- Already known or easily googleable (e.g., "Is the Earth round?")
- Binary sports/election results with near-term resolution dates and available oracles

Respond in JSON: {"valid": true/false, "reason": "brief explanation in English"}`
          },
          {
            role: "user",
            content: `Is this question suitable for a self-resolving prediction market?\n\nQuestion: "${question.trim()}"`
          }
        ],
        response_format: { type: "json_object" },
        temperature: 0.3,
      });

      const result = JSON.parse(response.choices[0].message.content);
      res.json({ valid: !!result.valid, reason: result.reason || "" });
    } catch (e) {
      console.error(`[Validate] LLM error: ${e.message}`);
      res.json({ valid: true, reason: "Validation unavailable, proceeding." });
    }
  });

  app.get("/api/agent-names", (req, res) => {
    res.json(getAgentNameMap());
  });

  app.get("/api/protocol", async (req, res) => {
    const { client } = resolveClient(req);
    if (!client) return res.status(503).json({ detail: "Chain not initialized" });
    try {
      const config = await client.getProtocolConfig();
      res.json({
        owner: config.owner,
        treasury: config.treasury,
        protocol_fee_bps: config.protocolFeeBps,
        protocol_fee_percent: config.protocolFeeBps / 100,
      });
    } catch (e) {
      res.status(500).json({ detail: e.message });
    }
  });

  const server = app.listen(port, "0.0.0.0", () => {
    console.log(`[API] REST server running on http://0.0.0.0:${port}`);
    console.log(`[API] Supported chains: ${Object.keys(CHAINS).join(", ")}`);
    console.log(`[API] Usage: /api/markets?chain=base or /api/markets?chain=monad`);
  });

  return server;
}

File: agents/api_server.py
"""
Yiling Protocol - REST API Server
Provides endpoints for external agent registration and market data queries.

Run standalone:  uvicorn api_server:app --host 0.0.0.0 --port 8000
Or import start_api_server() from run.py to run alongside the agent system.
"""

import threading
import uvicorn
from fastapi import FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel, HttpUrl

from typing import Optional
from agent_registry import AgentRegistry

# ─── App & Registry ───────────────────────────────────────────────────────────

app = FastAPI(
    title="Yiling Protocol API",
    description="Oracle-Free Self-Resolving Prediction Market — Agent SDK API",
    version="1.0.0",
)

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_methods=["*"],
    allow_headers=["*"],
)

registry = AgentRegistry()

# These get injected at runtime by run.py via set_shared_state()
_market_client = None
_orchestrator = None
_chain_watcher = None


def set_shared_state(market_client, orchestrator, chain_watcher=None):
    """Called by run.py to inject shared market_client, orchestrator, and/or chain_watcher."""
    global _market_client, _orchestrator, _chain_watcher
    _market_client = market_client
    _orchestrator = orchestrator
    _chain_watcher = chain_watcher


# ─── Request / Response Models ─────────────────────────────────────────────────

class RegisterAgentRequest(BaseModel):
    name: str
    webhook_url: HttpUrl
    wallet_address: str
    description: str = ""


class RegisterAgentResponse(BaseModel):
    id: str
    name: str
    api_key: str
    message: str


# ─── Agent Endpoints ───────────────────────────────────────────────────────────

@app.post("/api/agents/register", response_model=RegisterAgentResponse)
def register_agent(req: RegisterAgentRequest):
    """Register a new external prediction agent."""
    try:
        record = registry.register(
            name=req.name,
            webhook_url=str(req.webhook_url),
            wallet_address=req.wallet_address,
            description=req.description,
        )
    except ValueError as e:
        raise HTTPException(status_code=409, detail=str(e))

    return RegisterAgentResponse(
        id=record["id"],
        name=record["name"],
        api_key=record["api_key"],
        message=f"Agent '{record['name']}' registered successfully. Save your api_key — it cannot be retrieved later.",
    )


@app.delete("/api/agents/{agent_id}")
def unregister_agent(agent_id: str):
    """Remove a registered agent."""
    if not registry.unregister(agent_id):
        raise HTTPException(status_code=404, detail="Agent not found")
    return {"message": "Agent unregistered successfully"}


@app.get("/api/agents")
def list_agents():
    """List all registered agents (api_key is redacted)."""
    agents = registry.list_all()
    # Redact api_key for security
    return [
        {
            "id": a["id"],
            "name": a["name"],
            "webhook_url": a["webhook_url"],
            "wallet_address": a["wallet_address"],
            "description": a["description"],
            "predictions_made": a["predictions_made"],
            "last_active": a["last_active"],
            "active": a["active"],
            "registered_at": a["registered_at"],
        }
        for a in agents
    ]


# ─── Market Endpoints ─────────────────────────────────────────────────────────

@app.get("/api/markets")
def list_markets():
    """List all markets with basic info."""
    if not _market_client:
        raise HTTPException(status_code=503, detail="Market client not initialized")

    count = _market_client.get_market_count()
    markets = []
    for i in range(count):
        try:
            info = _market_client.get_market_info(i)
            is_active = _market_client.is_market_active(i) if not info["resolved"] else False
            markets.append({
                "market_id": i,
                "question": info["question"],
                "creator": info["creator"],
                "current_price": info["current_price"] / 1e18,
                "resolved": info["resolved"],
                "is_active": is_active,
                "prediction_count": info["prediction_count"],
                "total_pool": info["total_pool"] / 1e18,
            })
        except Exception:
            pass
    return markets


@app.get("/api/markets/count")
def get_market_count():
    """Get total number of markets created."""
    if not _market_client:
        raise HTTPException(status_code=503, detail="Market client not initialized")

    return {"count": _market_client.get_market_count()}


@app.get("/api/markets/{market_id}")
def get_market(market_id: int):
    """Get detailed info for a single market including prediction history."""
    if not _market_client:
        raise HTTPException(status_code=503, detail="Market client not initialized")

    try:
        info = _market_client.get_market_info(market_id)
        params = _market_client.get_market_params(market_id)
        predictions = _market_client.get_predictions(market_id)
    except Exception as e:
        raise HTTPException(status_code=404, detail=f"Market not found: {e}")

    return {
        "market_id": market_id,
        "question": info["question"],
        "current_price": info["current_price"] / 1e18,
        "creator": info["creator"],
        "resolved": info["resolved"],
        "total_pool": info["total_pool"] / 1e18,
        "prediction_count": info["prediction_count"],
        "params": {
            "alpha": params["alpha"] / 1e18,
            "k": params["k"],
            "flat_reward": params["flat_reward"] / 1e18,
            "bond_amount": params["bond_amount"] / 1e18,
            "liquidity_param": params["liquidity_param"] / 1e18,
            "created_at": params["created_at"],
        },
        "predictions": [
            {
                "index": i,
                "predictor": p["predictor"],
                "probability": p["probability"] / 1e18,
                "price_before": p["price_before"] / 1e18,
                "price_after": p["price_after"] / 1e18,
                "bond": p["bond"] / 1e18,
                "timestamp": p["timestamp"],
            }
            for i, p in enumerate(predictions)
        ],
    }


# ─── Leaderboard Endpoint ─────────────────────────────────────────────────────

@app.get("/api/leaderboard")
def get_leaderboard():
    """Get current agent leaderboard rankings.
    Uses orchestrator leaderboard if available, otherwise computes from on-chain data."""
    if _orchestrator:
        # Orchestrate mode: use in-memory leaderboard
        sorted_lb = sorted(
            _orchestrator.leaderboard.items(),
            key=lambda x: x[1],
            reverse=True,
        )
        return {
            "rankings": [
                {"rank": i + 1, "agent": name, "total_mon": round(mon, 6)}
                for i, (name, mon) in enumerate(sorted_lb)
            ]
        }

    if _chain_watcher:
        # Watch mode: compute leaderboard from on-chain data
        leaderboard = _chain_watcher.get_leaderboard()
        sorted_lb = sorted(leaderboard.items(), key=lambda x: x[1], reverse=True)
        return {
            "rankings": [
                {"rank": i + 1, "agent": name, "total_mon": round(mon, 6)}
                for i, (name, mon) in enumerate(sorted_lb)
            ]
        }

    if not _market_client:
        raise HTTPException(status_code=503, detail="No data source available")

    # Fallback: compute directly from chain (expensive)
    leaderboard = _compute_onchain_leaderboard()
    sorted_lb = sorted(leaderboard.items(), key=lambda x: x[1], reverse=True)
    return {
        "rankings": [
            {"rank": i + 1, "agent": name, "total_mon": round(mon, 6)}
            for i, (name, mon) in enumerate(sorted_lb)
        ]
    }


def _compute_onchain_leaderboard():
    """Compute leaderboard directly from on-chain data (fallback)."""
    leaderboard = {}
    try:
        count = _market_client.get_market_count()
    except Exception:
        return leaderboard

    for market_id in range(count):
        try:
            info = _market_client.get_market_info(market_id)
            if not info["resolved"]:
                continue

            params = _market_client.get_market_params(market_id)
            bond = params["bond_amount"]
            pred_count = info["prediction_count"]

            seen = set()
            for idx in range(pred_count):
                pred = _market_client.get_prediction(market_id, idx)
                address = pred["predictor"]
                if address in seen:
                    continue
                seen.add(address)

                payout = _market_client.get_payout(market_id, address)
                net_mon = (payout - bond) / 1e18 if payout > 0 else -bond / 1e18
                short_addr = f"{address[:6]}...{address[-4:]}"
                leaderboard[short_addr] = leaderboard.get(short_addr, 0.0) + net_mon
        except Exception:
            continue

    return leaderboard


# ─── Resolution & Payouts ────────────────────────────────────────────────────

@app.get("/api/markets/{market_id}/resolution")
def get_market_resolution(market_id: int):
    """Get resolution status and payout info for a market."""
    if not _market_client:
        raise HTTPException(status_code=503, detail="Market client not initialized")

    try:
        info = _market_client.get_market_info(market_id)
        params = _market_client.get_market_params(market_id)
        bond_stats = _market_client.contract.functions.getMarketBondStats(market_id).call()
    except Exception as e:
        raise HTTPException(status_code=404, detail=f"Market not found: {e}")

    return {
        "market_id": market_id,
        "resolved": info["resolved"],
        "final_price": info["current_price"] / 1e18,
        "prediction_count": info["prediction_count"],
        "total_pool": info["total_pool"] / 1e18,
        "total_bonds": bond_stats[0] / 1e18,
        "total_payouts_allocated": bond_stats[1] / 1e18,
        "alpha": params["alpha"] / 1e18,
        "k": params["k"],
        "resolution_mechanism": "Random stop (alpha) or force-resolve after 2 days",
    }


@app.get("/api/markets/{market_id}/leaderboard")
def get_market_leaderboard(market_id: int):
    """Get per-market leaderboard rankings in a single call."""
    if not _market_client:
        raise HTTPException(status_code=503, detail="Market client not initialized")

    try:
        info = _market_client.get_market_info(market_id)
        params = _market_client.get_market_params(market_id)
        predictions = _market_client.get_predictions(market_id)
    except Exception as e:
        raise HTTPException(status_code=404, detail=f"Market not found: {e}")

    if not info["resolved"] or not predictions:
        return {"market_id": market_id, "resolved": info["resolved"], "rankings": []}

    # Get agent name mapping
    agent_name_map = _get_agent_name_map()
    bond = params["bond_amount"]

    # Fetch payout for each unique predictor
    seen = set()
    entries = []
    for pred in predictions:
        addr = pred["predictor"]
        if addr in seen:
            continue
        seen.add(addr)
        try:
            payout = _market_client.get_payout(market_id, addr)
            net_mon = (payout - bond) / 1e18 if payout > 0 else -bond / 1e18
            name = agent_name_map.get(addr.lower(), f"{addr[:6]}...{addr[-4:]}")
            entries.append({"agent": name, "total_mon": round(net_mon, 6)})
        except Exception:
            continue

    entries.sort(key=lambda x: x["total_mon"], reverse=True)
    return {"market_id": market_id, "resolved": True, "rankings": entries}


def _get_agent_name_map():
    """Get agent address -> name mapping."""
    import os
    from web3 import Account
    from agents.profiles import AGENT_PROFILES

    mapping = {}
    for i, profile in enumerate(AGENT_PROFILES):
        key = os.getenv(f"AGENT_KEY_{i+1}", "")
        if key:
            try:
                addr = Account.from_key(key).address
                mapping[addr.lower()] = profile["name"]
            except Exception:
                pass
    return mapping


@app.get("/api/markets/{market_id}/payouts/{address}")
def get_payout(market_id: int, address: str):
    """Get payout amount for a specific address in a resolved market."""
    if not _market_client:
        raise HTTPException(status_code=503, detail="Market client not initialized")

    try:
        payout = _market_client.get_payout(market_id, address)
        info = _market_client.get_market_info(market_id)
        has_claimed = _market_client.has_claimed(market_id, address)
        has_predicted = _market_client.has_predicted(market_id, address)
    except Exception as e:
        raise HTTPException(status_code=404, detail=f"Not found: {e}")

    return {
        "market_id": market_id,
        "address": address,
        "has_predicted": has_predicted,
        "payout_amount": payout / 1e18,
        "resolved": info["resolved"],
        "has_claimed": has_claimed,
    }


@app.get("/api/markets/{market_id}/predictions")
def get_predictions(market_id: int, index: Optional[int] = None):
    """Get predictions for a market. Pass ?index=N for a single prediction."""
    if not _market_client:
        raise HTTPException(status_code=503, detail="Market client not initialized")

    try:
        if index is not None:
            p = _market_client.get_prediction(market_id, index)
            return {
                "market_id": market_id,
                "index": index,
                "predictor": p["predictor"],
                "probability": p["probability"] / 1e18,
                "price_before": p["price_before"] / 1e18,
                "price_after": p["price_after"] / 1e18,
                "bond": p["bond"] / 1e18,
                "timestamp": p["timestamp"],
            }

        predictions = _market_client.get_predictions(market_id)
        return {
            "market_id": market_id,
            "count": len(predictions),
            "predictions": [
                {
                    "index": i,
                    "predictor": p["predictor"],
                    "probability": p["probability"] / 1e18,
                    "price_before": p["price_before"] / 1e18,
                    "price_after": p["price_after"] / 1e18,
                    "bond": p["bond"] / 1e18,
                    "timestamp": p["timestamp"],
                }
                for i, p in enumerate(predictions)
            ],
        }
    except Exception as e:
        raise HTTPException(status_code=404, detail=f"Not found: {e}")


# ─── Protocol Config ────────────────────────────────────────────────────────

@app.get("/api/protocol")
def get_protocol_config():
    """Get protocol-level configuration (owner, treasury, fee)."""
    if not _market_client:
        raise HTTPException(status_code=503, detail="Market client not initialized")

    try:
        config = _market_client.get_protocol_config()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to read config: {e}")

    return {
        "owner": config["owner"],
        "treasury": config["treasury"],
        "protocol_fee_bps": config["protocol_fee_bps"],
        "protocol_fee_percent": config["protocol_fee_bps"] / 100,
    }


@app.get("/api/markets/{market_id}/active")
def is_market_active(market_id: int):
    """Check if a market is still active (not yet resolved)."""
    if not _market_client:
        raise HTTPException(status_code=503, detail="Market client not initialized")

    try:
        active = _market_client.is_market_active(market_id)
        return {"market_id": market_id, "active": active}
    except Exception as e:
        raise HTTPException(status_code=404, detail=f"Market not found: {e}")


# ─── Agent Address Mapping ─────────────────────────────────────────────────────

@app.get("/api/agent-names")
def get_agent_names():
    """Get mapping of agent wallet addresses to names."""
    import os
    from web3 import Account
    from agents.profiles import AGENT_PROFILES

    mapping = {}
    for i, profile in enumerate(AGENT_PROFILES):
        key = os.getenv(f"AGENT_KEY_{i+1}", "")
        if key:
            try:
                addr = Account.from_key(key).address
                mapping[addr.lower()] = profile["name"]
            except Exception:
                pass
    return mapping


# ─── ABI Download ─────────────────────────────────────────────────────────────

@app.get("/api/abi")
def get_abi():
    """Download the contract ABI for direct on-chain interaction."""
    from fastapi.responses import JSONResponse
    from config import load_abi
    try:
        abi = load_abi()
        # Always return as a JSON array (not wrapped in an object)
        if isinstance(abi, list):
            return JSONResponse(content=abi)
        if isinstance(abi, dict) and "abi" in abi:
            return JSONResponse(content=abi["abi"])
        return JSONResponse(content=abi)
    except FileNotFoundError as e:
        raise HTTPException(status_code=500, detail=str(e))


# ─── Health ────────────────────────────────────────────────────────────────────

@app.get("/api/health")
def health():
    mode = "orchestrate" if _orchestrator else ("watch" if _chain_watcher else "standalone")
    return {
        "status": "ok",
        "mode": mode,
        "registered_agents": len(registry.list_all()),
        "market_client": _market_client is not None,
        "orchestrator": _orchestrator is not None,
        "chain_watcher": _chain_watcher is not None,
    }


# ─── Server Launcher ──────────────────────────────────────────────────────────

def start_api_server(host: str = "0.0.0.0", port: int = 8000):
    """Start the API server in a background thread (non-blocking)."""
    thread = threading.Thread(
        target=uvicorn.run,
        args=(app,),
        kwargs={"host": host, "port": port, "log_level": "warning"},
        daemon=True,
    )
    thread.start()
    print(f"[API] REST server running on http://{host}:{port}")
    print(f"[API] Docs: http://localhost:{port}/docs")
    return thread
